pytorch_lightning/loggers/neptune.py

```diff
@@ -46,247 +93,469 @@ class NeptuneLogger(LightningLoggerBase):
+                    You can still integrate it with the Neptune logger using legacy Python API
+                    available as part of neptune-contrib package:
+                      - https://docs-legacy.neptune.ai/integrations/pytorch_lightning.html\n
+                    """
+                ) from e
+
+        if not self._run_instance_initialized:
+            # make sure that we've log integration version for both newly created and outside `Run` instances
+            self._run_instance_initialized = True
+            self._run_instance[INTEGRATION_VERSION_KEY] = __version__
+
+        return self._run_instance
 
     @rank_zero_only
-    def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
-        params = self._convert_params(params)
-        params = self._flatten_dict(params)
-        for key, val in params.items():
-            self.experiment.set_property(f"param__{key}", val)
+    def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:  # skipcq: PYL-W0221
+        r"""
+        Log hyper-parameters to the run.
 
-    @rank_zero_only
-    def log_metrics(self, metrics: Dict[str, Union[torch.Tensor, float]], step: Optional[int] = None) -> None:
-        """
-        Log metrics (numeric values) in Neptune experiments.
+        Hyperparams will be logged under the "<prefix>/hyperparams" namespace.
+
+        **Note**
+
+        You can also log parameters by directly using the logger instance:
+        ``neptune_logger.experiment["model/hyper-parameters"] = params_dict``.
+
+        In this way you can keep hierarchical structure of the parameters.
 
         Args:
-            metrics: Dictionary with metric names as keys and measured quantities as values
-            step: Step number at which the metrics should be recorded, currently ignored
+            params: `dict`.
+                Python dictionary structure with parameters.
+
+        Example::
+
+            from pytorch_lightning.loggers import NeptuneLogger
+
+            PARAMS = {
+                "batch_size": 64,
+                "lr": 0.07,
+                "decay_factor": 0.97
+            }
+
+            neptune_logger = NeptuneLogger(
+                api_key="ANONYMOUS",
+                project="common/pytorch-lightning-integration"
+            )
+
+            neptune_logger.log_hyperparams(PARAMS)
         """
-        assert rank_zero_only.rank == 0, "experiment tried to log from global_rank != 0"
+        params = self._convert_params(params)
+        params = self._sanitize_callable_params(params)
+
+        parameters_key = self.PARAMETERS_KEY
+        parameters_key = self._construct_path_with_prefix(parameters_key)
+
+        self.run[parameters_key] = params
+
+    @rank_zero_only
+    def log_metrics(self, metrics: Dict[str, Union[torch.Tensor, float]], step: Optional[int] = None) -> None:
+        if rank_zero_only.rank != 0:
+            raise ValueError("run tried to log from global_rank != 0")
```

Comment:
there is a decorator here preventing this, at least in lightning.
did you add the error handling here expecting users to call this method manually?

Reply:
Yes I think it's redundant check.

Reply:
I've just made assertion here, but checkstyle forced me to change it to raising an Exception.

It's here because when something goes wrong and we'll try to log to the same logger from several threads, `neptune-client` may crash due to non synchronized timestamps sent to the series.
It's very unlikely that this happens because of mentioned `@rank_zero_only` but I've used this defensive strategy since it's much more easier to track such error from exception rather than mentioned problem with logging.